File: java_src/foam/core/AbstractFObject.java
package foam.core;

import java.util.List;
import java.util.Iterator;

public abstract class AbstractFObject
  extends ContextAwareSupport
  implements FObject
{

  public static FObject maybeClone(FObject fo) {
    return (fo == null ? null : ((AbstractFObject)fo).fclone());
  }

  public FObject fclone() {
    FObject ret = (FObject) getClassInfo().newInstance();
    if(ret != null) {
      List<PropertyInfo> props = getClassInfo().getAxiomsByClass(PropertyInfo.class);
      for( PropertyInfo pi : props ) {
        pi.set(ret, pi.get(this));
      }
    }
    return ret;
  }

  public int compareTo(Object o) {
    if ( o == this ) return 0;
    if ( o == null ) return 1;
    if ( ! ( o instanceof FObject ) ) return 1;

    if ( getClass() != o.getClass() ) {
      return getClassInfo().getId().compareTo(((FObject)o).getClassInfo().getId());
    }

    List props = getClassInfo().getAxiomsByClass(PropertyInfo.class);
    Iterator i = props.iterator();

    int result;
    while ( i.hasNext() ) {
      result = ((PropertyInfo)i.next()).compare(this, o);
      if ( result != 0 ) return result;
    }

    return 0;
  }
}

File: src/foam/core/AbstractFObject.js
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.core',
  name: 'AbstractFObject',
  abstract: true,
  flags: ['java'],

  javaImports: [
    'foam.lib.json.Outputter',
    'foam.util.SecurityUtil',
    'java.security.*',
    'java.util.HashMap',
    'java.util.Iterator',
    'java.util.List',
    'java.util.Map'
  ],

  methods: [
    {
      name: 'hashCode',
      type: 'Integer',
      javaCode: `
        int hashCode = 1;
        List props = getClassInfo().getAxiomsByClass(PropertyInfo.class);
        Iterator i = props.iterator();

        while ( i.hasNext() ) {
          PropertyInfo pi = (PropertyInfo) i.next();
          hashCode = 31 * hashCode + java.util.Objects.hash(pi.get(this));
        }

        return hashCode;
      `
    },
    {
      name: 'equals',
      type: 'Boolean',
      args: [ { name: 'o', type: 'Any' } ],
      javaCode: `
        return compareTo(o) == 0;
      `
    }
  ]
});
